### PR TITLE
codec: pre-alloc encode buffer in row level

### LINKED ISSUE #11202: codec: Pre-realloc encode row buffer
## Performance Questions

- What version of TiDB are you using?

```
3.0.0 GA
```

- What's the observed and your expected performance respectively?

```
tablecodec.EncodeRow 
````
https://github.com/pingcap/tidb/blob/2106f6198bf08e5e823b1420056cde5e22526d92/tablecodec/tablecodec.go#L227

will encode a row data into a row byte.

https://github.com/pingcap/tidb/pull/5600 introduce optimize to reuse encode buffer.

but some other place pass `valBuf == nil`

- binlog (update/delete...) it effect when open binlog
- batch_checker when  we using replace/ on duplicate

regardless of reuse problem.... current problem is:

encode buffer will realloc/grow for *each columns*.

for example, we have a  table with 200+ varchar columns, current logic will check and realloc for each datums at 

https://github.com/pingcap/tidb/blob/2729df8e8dc198aec17ab388a68a082520df79ab/util/codec/bytes.go#L149

so it maybe will realloc for each columns(200 times for a row), and later columns realloc bigger..

cpu of Replace.Next

image

heap alloc-spaces of all

image

we can try let the buffer be reusable, and pre-calculate and realloc in per-row instead of per-column

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

the first part to fix  #11202

### What is changed and how it works?

pre-calculate buffer and realloc once

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - N/A

Side effects

 - N/A

Related changes

 - N/A

Benchmark

benchmark with 178 columns table + replace into test(this test doesn't use prepare and plan cache if so will got better result)

before

```
[ 5s ] thds: 32 tps: 569.83 qps: 1146.04 (r/w/o: 576.21/569.83/0.00) lat (ms,95%): 106.75 err/s: 0.00 reconn/s: 0.00
[ 10s ] thds: 32 tps: 596.25 qps: 1192.50 (r/w/o: 596.25/596.25/0.00) lat (ms,95%): 95.81 err/s: 0.00 reconn/s: 0.00
[ 15s ] thds: 32 tps: 610.01 qps: 1219.42 (r/w/o: 609.41/610.01/0.00) lat (ms,95%): 97.55 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 32 tps: 599.60 qps: 1198.40 (r/w/o: 598.80/599.60/0.00) lat (ms,95%): 101.13 err/s: 0.00 reconn/s: 0.00
[ 25s ] thds: 32 tps: 588.16 qps: 1177.73 (r/w/o: 589.56/588.16/0.00) lat (ms,95%): 102.97 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 32 tps: 613.64 qps: 1226.89 (r/w/o: 613.24/613.64/0.00) lat (ms,95%): 99.33 err/s: 0.00 reconn/s: 0.00
[ 35s ] thds: 32 tps: 606.35 qps: 1212.50 (r/w/o: 606.15/606.35/0.00) lat (ms,95%): 106.75 err/s: 0.00 reconn/s: 0.00
```

after

```
[ 5s ] thds: 32 tps: 752.44 qps: 1510.67 (r/w/o: 758.23/752.44/0.00) lat (ms,95%): 90.78 err/s: 0.00 reconn/s: 0.00
[ 10s ] thds: 32 tps: 712.86 qps: 1425.32 (r/w/o: 712.46/712.86/0.00) lat (ms,95%): 82.96 err/s: 0.00 reconn/s: 0.00
[ 15s ] thds: 32 tps: 750.60 qps: 1501.21 (r/w/o: 750.60/750.60/0.00) lat (ms,95%): 87.56 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 32 tps: 728.80 qps: 1458.00 (r/w/o: 729.20/728.80/0.00) lat (ms,95%): 90.78 err/s: 0.00 reconn/s: 0.00
[ 25s ] thds: 32 tps: 707.99 qps: 1416.39 (r/w/o: 708.39/707.99/0.00) lat (ms,95%): 89.16 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 32 tps: 746.20 qps: 1492.60 (r/w/o: 746.40/746.20/0.00) lat (ms,95%): 84.47 err/s: 0.00 reconn/s: 0.00
[ 35s ] thds: 32 tps: 754.41 qps: 1508.61 (r/w/o: 754.21/754.41/0.00) lat (ms,95%): 86.00 err/s: 0.00 reconn/s: 0.00
```

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/11218)

### CODE REVIEW

File: util/codec/codec.go [hunk 1]
```diff
@@ -42,9 +42,41 @@ const (
...
+			return b
+		}
+	}
+	return reallocBytes(b, size)
+}
+
 // encode will encode a datum and append it to a byte slice. If comparable is true, the encoded bytes can be sorted as it's original order.
 // If hash is true, the encoded bytes can be checked equal as it's original value.
 func encode(sc *stmtctx.StatementContext, b []byte, vals []types.Datum, comparable bool, hash bool) (_ []byte, err error) {
+	if b == nil {
```
Comment (zz-jason) [inline util/codec/codec.go]: Does this optimization only reduce the `runtime.growslice()` overhead when the upper layer passes `nil` pointer to this function?

File: util/codec/decimal.go [hunk 2]
```diff
@@ -30,6 +30,13 @@ func EncodeDecimal(b []byte, dec *types.MyDecimal, precision, frac int) ([]byte,
 	return b, errors.Trace(err)
 }
 
+func sizeDecimal(dec *types.MyDecimal, precision, frac int) int {
```
Comment (zz-jason) [inline util/codec/decimal.go]: Can we use `types.MyDecimalStructSize` as the estimated value?

Comment (lysu) [reply util/codec/decimal.go]: this method is port from MySQL and it seems can return an accurate value - -, but it's very complex..it's a good idea to use estimated value.

File: util/codec/codec.go (hunk 1 above)
Comment (lysu) [reply util/codec/codec.go]: previous @coocood think we should avoid calculating cost when outer give us a buffer, but now we use estimated value for decimal so we can remove this null check? @coocood

File: util/codec/codec.go [hunk 3]
```diff
@@ -42,9 +42,33 @@ const (
 	maxFlag          byte = 250
 )
 
+func preRealloc(b []byte, vals []types.Datum, comparable bool, hash bool) []byte {
```
Comment (ngaut) [inline util/codec/codec.go]: The hash argument is not used, please remove it.

**Comment by lysu** [comment]
ping @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**